# Movie Theatre

AOC Day 9 Part 1

In [380]:
input = [
    "7,1",
    "11,1",
    "11,7",
    "9,7",
    "9,5",
    "2,5",
    "2,3",
    "7,3"
]

coordinates = [
    (int(inp.split(",")[0]), int(inp.split(",")[1])) for inp in input
]

coordinates

[(7, 1), (11, 1), (11, 7), (9, 7), (9, 5), (2, 5), (2, 3), (7, 3)]

In [381]:
max_area = 0

for I in range(len(coordinates) - 1):
    for J in range(1, len(coordinates)):
        iy, ix = coordinates[I]
        jy, jx = coordinates[J]
        if ix == jx:
            # consider the unit length of a row.
            max_area = max(abs(iy - jy + 1), max_area)
        elif iy == jy:
            # consider the unit length of a column.
            max_area = max(abs(ix - iy + 1), max_area)
        else:
            max_area = max(abs(ix - jx + 1) * abs(iy - jy + 1), max_area)

print(max_area)

50


$50$ is the answer!

Let's try it for `input.txt`.

In [382]:
with open(file="input.txt") as file:
    input = [line.rstrip() for line in file]

    coordinates = [
        (int(inp.split(",")[0]), int(inp.split(",")[1])) for inp in input
    ]       

    max_area = 0

    for I in range(len(coordinates) - 1):
        for J in range(1, len(coordinates)):
            iy, ix = coordinates[I]
            jy, jx = coordinates[J]
            if ix == jx:
                # consider the unit length of a row.
                max_area = max(abs(iy - jy + 1), max_area)
            elif iy == jy:
                # consider the unit length of a column.
                max_area = max(abs(ix - iy + 1), max_area)
            else:
                max_area = max(abs(ix - jx + 1) * abs(iy - jy + 1), max_area)

    print(max_area)

4781235324


$4781235324$ is the correct answer! 

That was super easy, so Part 2 must be the real deal.

---

### Part II

Let us take in our `Matrix` implementation here.

In [383]:
class Neighbours:
    def __init__(self, a: str | None, b: str | None, c: str | None, m: str | None, n: str | None, p: str | None, q: str | None, r: str | None) -> None:
        self.a = a
        self.b = b
        self.c = c
        self.m = m
        self.n = n
        self.p = p
        self.q = q
        self.r = r

    # This function returns what's immediately above the current node
    # i.e., the North direction.
    @staticmethod
    def parent(matrix: "Matrix", of: tuple[int, int]) -> str | None:
        return Neighbours.of(matrix=matrix, node=of).b

    @staticmethod
    def of(matrix: "Matrix", node: tuple[int, int]):
        # x = rows, y = cols
        (x, y) = node
        x_min = True if x - 1 < 0 else False
        y_min = True if y - 1 < 0 else False

        x_max = True if x + 1 >= matrix.num_rows() else False
        y_max = True if y + 1 >= matrix.num_cols() else False

        a = None if (x_min or y_min) else matrix[x-1, y-1]
        b = None if x_min else matrix[x -1, y]
        c = None if (x_min or y_max) else matrix[x - 1, y + 1]
        m = None if y_min else matrix[x, y - 1]
        p = None if (y_min or x_max) else matrix[x + 1, y - 1]
        n = None if y_max else matrix[x, y + 1]
        r = None if x_max else matrix[x + 1, y]
        q = None if (y_max or x_max) else matrix[x + 1, y + 1]

        return Neighbours(
            a, b, c, m, n, p, q, r
        )
    
    def to_dict(self):
        return {
            "a": self.a,
            "b": self.b,
            "c": self.c,
            "m": self.m,
            "n": self.n,
            "p": self.p,
            "q": self.q,
            "r": self.r
        }
    
    def __iter__(self):
        yield from [
            self.a,
            self.b,
            self.c,
            self.m,
            self.n,
            self.p,
            self.r,
            self.q
        ]

class Matrix:
    def __init__(self, rows: list[list[str]]) -> None:
        self.current_idx = -1
        self.last_idx = len(rows)
        self._vec: list[list] = []
        self._prepare_matrix(rows)

    def __iter__(self):
        return self
    
    def __next__(self):
        self.current_idx += 1
        if self.current_idx < self.last_idx:
            return self._vec[self.current_idx]

    def _prepare_matrix(self, rows: list[list[str]]) -> None:
        self._vec = [line for line in rows]

    def __getitem__(self, key: tuple[int, int]):
        """
        key: (row, column)
        """
        (row, col) = key
        return self._vec[row][col]

    def __setitem__(self, key: tuple[int, int], value: str):
        """
        key: (row, column)
        """
        (row, col) = key
        self._vec[row][col] = value

    def __len__(self):
        return len(self._vec)
    
    def neighbours(self, node: tuple[int, int]):
        return Neighbours.of(
            matrix=self,
            node=node
        )
    
    def row(self, idx: int) -> list[str]:
        return self._vec[idx]

    def num_rows(self) -> int:
        return len(self._vec)
    
    def num_cols(self) -> int:
        return len(self._vec[0])
    
    def to_string(self) -> str:
        return str(self._vec)
    
    def pretty_print(self) -> None:
        for row in self._vec:
            print("".join(row))
    
    def dim(self) -> tuple[int, int]:
        return (self.num_rows(), self.num_cols())
    
    def save(self, filename: str):
        """
        Save the matrix to a file, one row per line.
        Each row is joined into a string.
        """
        with open(filename, 'w') as f:
            for row in self._vec:
                f.write(''.join(row) + '\n')

Let's first aim to arrive at the following matrix.

```
..............
.......#XXX#..
.......X...X..
..#XXXX#...X..
..X........X..
..#XXXXXX#.X..
.........X.X..
.........#X#..
..............
```

In [384]:
input = [
    "7,1",
    "11,1",
    "11,7",
    "9,7",
    "9,5",
    "2,5",
    "2,3",
    "7,3"
]

coordinates = [
    (int(inp.split(",")[0]), int(inp.split(",")[1])) for inp in input
]

# Assume a 15 x 10 grid.
rows = [["."] * 15 for _ in range(10)]

matrix = Matrix(rows=rows)
matrix.pretty_print()

...............
...............
...............
...............
...............
...............
...............
...............
...............
...............


In [385]:
for y, x in coordinates:
    matrix[x, y] = "#"

matrix.pretty_print()

...............
.......#...#...
...............
..#....#.......
...............
..#......#.....
...............
.........#.#...
...............
...............


First, let us check rows.

In [386]:
coordinates

[(7, 1), (11, 1), (11, 7), (9, 7), (9, 5), (2, 5), (2, 3), (7, 3)]

In [387]:
full_coordinates = []

for I in range(len(coordinates) - 1):
    for J in range(1, len(coordinates)):
        iy, ix = coordinates[I]
        jy, jx = coordinates[J]
        # Same row 
        if ix == jx:
            # consider the unit length of a row.
            # We need to fill the distance with "X"s.
            d = abs(iy - jy) + 1
            s = min(iy, jy)
            for i in range(1, d - 1):
                matrix[ix, s + i] = "X"
                full_coordinates.append((s + i, ix))
        # Same column
        elif iy == jy:
            # consider the unit length of a column.
            d = abs(ix - jx) + 1
            s = min(ix, jx)
            for i in range(1, d - 1):
                matrix[s + i, iy] = "X"
                full_coordinates.append((iy, s + i))
        else:
            continue

for I in range(len(full_coordinates) - 1):
    for J in range(1, len(full_coordinates)):
        iy, ix = full_coordinates[I]
        jy, jx = full_coordinates[J]
        # Same row 
        if ix == jx:
            # consider the unit length of a row.
            # We need to fill the distance with "X"s.
            d = abs(iy - jy) + 1
            s = min(iy, jy)
            for i in range(1, d - 1):
                if matrix[ix, s + i] != "#":
                    matrix[ix, s + i] = "X"
        # Same column
        elif iy == jy:
            # consider the unit length of a column.
            d = abs(ix - jx) + 1
            s = min(ix, jx)
            for i in range(1, d - 1):
                if matrix[s + i, iy] != "#":
                    matrix[s + i, iy] = "X"
        else:
            continue

matrix.pretty_print()


...............
.......#XXX#...
.......XXXXX...
..#XXXX#XXXX...
..XXXXXXXXXX...
..#XXXXXX#XX...
.........XXX...
.........#X#...
...............
...............


In addition, all of the tiles inside this loop of red and green tiles need also to be green. That is, it needs to be like the following:

```
..............
.......#XXX#..
.......XXXXX..
..#XXXX#XXXX..
..XXXXXXXXXX..
..#XXXXXX#XX..
.........XXX..
.........#X#..
..............
```

Looks like we are good. We have a big time and space complexity problem but let's try to go ahead with what we have now.